### ISSUE TITLE
[Error] Monster walking behaviour

### ISSUE THREAD

**Issue by ThisMushroom** [MUnique/OpenMU#39]
### Overview:

1) Add new SpawnArea with any kind of monster (e.g. Giants (Lorencia))
2) Make SpawnArea as a single cell spot (x1 == x1 && y1 == y2)
3) Make SpawnArea spawning 20 monsters
4) Reinit database for spawn area initialization 
5) Walk to the spawn area 
6) Observe empty spot

### Observation:

1) Monsters are walking towards 0,0 coordinates of the map.
2) Some monsters remain standing/walking around the spot area.
3) For DirectHit spell system monsters that remain standing/walking are acting correctly when being hit.
3.1) If you want to hit them with DirectHit spell - hit them - they will die, you will get exp/zen. No drop.
4) For AreaHit spell system those monsters that remain standing/walking are invisible.
4.1) If you want to hit them with AreaHit spell - try to - the will act like being hit but no damage would be done. Therefore as we can see from debugging - system doesn't detect monsters around you in the specified skill range. 

### Ideas:

1) Yup, monsters are walking thats obvious. - Need to implement correct work of walk range (like the definition of how far they can move from initial spawn area (for the single cell spawns) and new Point distance (for area spot that is bigger then walk range). Moreover implement logic to prevent monsters from leaving the SpawnArea.

2(+ 3-4)) Didn't debug it well yet, just shortlooked. But first ideas and observations were: 
- If you take Evil Spirits and start casting somewhere in the corner of the map near 0,0 (as close as you can) - you will hit those monsters (but you cannot see them). Why do I think they stand at 0,0 - because of EXP trail from monster to caster after being killed (not alway working) + you get exp and zen.
- It may be problem with server-client sync and maybe somehow bucket system bug. 
When on server side monsters walked to 0,0 but on client side they are displayed on other places. (take a look at the map in the admin panel - all monsters are on correct places but ingame - they are not).
Why we can hit them with DirectHit spells - because client send the data about monster and it is ok to calculate hit.
Why we cannot hit them with AreaHit spell - because non-target spell use being sent to the server and the server should determine monsters being hit int the spell range around the caster.

**PS. I would investigate it deeper this eve and add comments about to make it more detailed.**

**Comment by sven-n** [comment]
Thanks again 👍 You're probably the first one who tried to learn and perform a skill on the OpenMU server 🥇 
I wondered that it even worked somehow, btw :)

Just a wild guess:
The walking to 0,0 could be an issue in the packets (WorldView.NewNpcsInScope). There is a start and end coordinate in the meet packet, which are different when the monster walks. Maybe the walk target is not initialized yet here, or IsWalking is not correctly assigned:

    ```csharp
                if (supportWalk?.IsWalking ?? false)
                {
                    packet[11 + monsterOffset] = supportWalk.WalkTarget.X;
                    packet[12 + monsterOffset] = supportWalk.WalkTarget.Y;
                }
                else
                {
                    packet[11 + monsterOffset] = npc.X;
                    packet[12 + monsterOffset] = npc.Y;
                }
    ```

The evil spirit skill could cause some area skill hit packets for these monsters at 0,0 when they're visible (or in range) on client side. Currently there is no check on the server yet to prevent such hits.

About the walking in general: Monsters are allowed to walk outside of their spawn area. At the original game (global server) it was possible to lure monsters across the whole map.

**Comment by sven-n** [comment]
I made some fixes for area skills... they should find the monsters in range now - provided they are inside a bucket which is in range. Time to sleep now 😴

**Comment by sven-n** [comment]
I found the issue... As soon as a Monster is doing a "RandomMove", it moves towards 0,0. The reason is, that RandomMove doesn't set the WalkTarget.

**Comment by ThisMushroom** [comment]
Another thing "not so far from moving" is monster respawn. Seems to be packet isssue.
If one create a small spawnarea (or point spawnarea for better example) -> as soon as you kinn those monsters you will not see new spawned monsters but on the map at the admin panel you can see them.

**Comment by sven-n** [comment]
Fixed the respawn issue. However, I noticed something related again. When monsters attack, they don't seem to be in the correct coordinate or looking at the right direction.

**Comment by ThisMushroom** [comment]
Tested the fix. 
At the first look - fix works. But over time - same issue occure. There are simply no visible mobs.
If we assume that map in admin panel works perfect - _next thing become interesting:_ "invisible mobs" are not moving at all. (Just newly respawned mob dot appears but they are not moving).

**Comment by sven-n** [comment]
Oh, that's strange...
I can only imagine it's related to walking. Can you test if this issue occurs when they don't walk at all, by deactivating the walk timer and random move?

Anyways, thanks for the test :)

**Comment by sven-n** [comment]
I think this issue has been fixed with #47.


> However, I noticed something related again. When monsters attack, they don't seem to be in the correct coordinate or looking at the right direction.

I'll create an issue, if that's still occuring.
